File: skeleton.c
/* skeleton.c - typical starting point for how I like to write command-line
 * tools. See https://github.com/fordsfords/skeleton
 * This tries to be portable between Mac, Linux, and Windows.
 */
/*
# This code and its documentation is Copyright 2002-2021 Steven Ford
# and licensed "public domain" style under Creative Commons "CC0":
#   http://creativecommons.org/publicdomain/zero/1.0/
# To the extent possible under law, the contributors to this project have
# waived all copyright and related or neighboring rights to this work.
# In other words, you can use this code for any purpose without any
# restrictions.  This work is published from: United States.  The project home
# is https://github.com/fordsfords/skeleton
*/

#include <stdio.h>
#include <string.h>
#include <time.h>
#include <errno.h>
#ifdef _WIN32
  #include <winsock2.h>
  #define SLEEP(s) Sleep((s)*1000)
#else
  #include <stdlib.h>
  #include <unistd.h>
  #define SLEEP(s) sleep(s)
#endif

#include "skeleton.h"


/* Options and their defaults */
int o_testnum = 0;


char usage_str[] = "Usage: skeleton [-h] [-t testnum]";

void usage(char *msg) {
  if (msg) fprintf(stderr, "%s\n", msg);
  fprintf(stderr, "%s\n", usage_str);
  exit(1);
}

void help() {
  fprintf(stderr, "%s\n", usage_str);
  fprintf(stderr, "where:\n"
                  "  -h : print help\n"
                  "  -t testnum : run specified test\n");
  exit(0);
}


int main(int argc, char **argv)
{
  int opt;

#ifdef _WIN32
  /* windows-specific code */
  WSADATA wsadata;
  int wsStat = WSAStartup(MAKEWORD(2,2), &wsadata);
  if (wsStat != 0) {printf("line %d: wsStat=%d\n",__LINE__,wsStat);exit(1);}
#endif

  while ((opt = getopt(argc, argv, "ht:")) != EOF) {
    switch (opt) {
      case 't':
        SAFE_ATOI(optarg, o_testnum);
        break;
      case 'h':
        help();
        break;
      default:
        usage(NULL);
    }  /* switch opt */
  }  /* while getopt */

  if (optind != argc) { usage("Extra parameter(s)"); }

  switch(o_testnum) {
    case 0:  /* no test */
      break;

    case 1:
      fprintf(stderr, "ASSRT\n");
      ASSRT(o_testnum == 1 && "internal test fail");
      ASSRT(o_testnum != 1 && "should fail");
      break;

    case 2:
    {
      FILE *perr_fp;
      fprintf(stderr, "PERR\n");
      perr_fp = fopen("file_not_exist", "r");
      if (perr_fp == NULL) {
        PERR("errno should be 'file not found':");
      } else {
        ABRT("Internal test failure: 'file_not_exist' appears to exist");
      }
      break;
    }

    case 3:
    {
      FILE *perr_fp;
      fprintf(stderr, "EOK0\n");
      perr_fp = fopen("skeleton.c", "r");
      if (perr_fp == NULL) {
        ABRT("Internal test failure: 'skeleton.c' appears to not exist");
      } else {
        EOK0(fclose(perr_fp) && "internal test fail");
        EOK0(fclose(perr_fp) && "should fail with bad file descr");
      }
      break;
    }

    case 4:
    {
      FILE *perr_fp;
      fprintf(stderr, "ENULL\n");
      ENULL(perr_fp = fopen("skeleton.c", "r")); /* should be OK. */
      ENULL(perr_fp = fopen("file_not_exist", "r")); /* should fail. */
      break;
    }

    case 5:
      fprintf(stderr, "ABRT\n");
      ABRT("ABRT test");
      break;

    case 6:
      fprintf(stderr, "SLEEP_SEC\n");
      SLEEP_SEC(1);
      fprintf(stderr, "Done\n");
      break;

    case 7:
      fprintf(stderr, "SLEEP_MS 1000\n");
      SLEEP_MS(1000);
      fprintf(stderr, "Done\n");
      break;

    case 8:
    {
      char *str, *word, *context;

      fprintf(stderr, "STRTOK_PORT\n");
      str = strdup("abc,xyz,123");
      for (word = strtok_r(str, ",", &context);
          word != NULL;
          word = strtok_r(NULL, ",", &context)) {
        printf("word='%s'\n", word);
      }

      break;
    }

    default: /* ABRT */
      ABRT("unknown option, aborting.");
  }

#ifdef _WIN32
  WSACleanup();
#endif
}  /* main */

File: skeleton.pl
#!/usr/bin/env perl
# skeleton.pl

# This work is dedicated to the public domain under CC0 1.0 Universal:
# http://creativecommons.org/publicdomain/zero/1.0/
# 
# To the extent possible under law, Steven Ford has waived all copyright
# and related or neighboring rights to this work. In other words, you can 
# use this code for any purpose without any restrictions.
# This work is published from: United States.
# Project home: https://github.com/fordsfords/skeleton

use strict;
use warnings;
use Getopt::Std;
use File::Basename;
use Carp;

# globals
my $tool = basename($0);

# process options.
use vars qw($opt_h $opt_o);
getopts('ho:') || mycroak("getopts failure");

if (defined($opt_h)) {
  help();
}

my $out_fd;
if (defined($opt_o)) {
  open($out_fd, ">", $opt_o) or mycroak("Error opening '$opt_o': $!");
} else {
  $out_fd = *STDOUT;
}

# Main loop; read each line in each file.
while (<>) {
  chomp;  # remove trailing \n

  # glue continuation lines and strip comments, leading/trailing spaces, and blank lines.
  if (s/\\$//) {
    $_ .= <>;
    redo unless eof();
  }
  s/#.*$//;  s/\s+$//;  s/^\s+//;
  next if (/^\s*$/);

  # do rest of work.
  print $out_fd "??? File: $ARGV, line: $.: '$_'\n";
} continue {  # This continue clause makes "$." give line number within file.
  close ARGV if eof;
}

# All done.
exit(0);


# End of main program, start subroutines.


sub mycroak {
  my ($msg) = @_;

  if (defined($ARGV)) {
    # Print input file name and line number.
    croak("Error (use -h for help): input_file:line=$ARGV:$., $msg");
  } else {
    croak("Error (use -h for help): $msg");
  }
}  # mycroak


sub assrt {
  my ($assertion, $msg) = @_;

  if (! ($assertion)) {
    if (defined($msg)) {
      mycroak("Assertion failed, $msg");
    } else {
      mycroak("Assertion failed");
    }
  }
}  # assrt


sub help {
  my($err_str) = @_;

  if (defined $err_str) {
    print "$tool: $err_str\n\n";
  }
  print <<__EOF__;
Usage: $tool [-h] [-o out_file] [file ...]
Where ('R' indicates required option):
    -h - help
    -o out_file - output file (default: STDOUT).
    file ... - zero or more input files.  If omitted, inputs from stdin.

__EOF__

  exit(0);
}  # help
